subsys/usb/usb_device.c

```diff
@@ -1004,3 +1040,234 @@ int usb_ep_read_continue(u8_t ep)
 {
 	return usb_dc_ep_read_continue(ep);
 }
+
+/* Transfer management */
+static struct usb_transfer_data *usb_ep_get_transfer(u8_t ep)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(usb_dev.transfer); i++) {
+		if (usb_dev.transfer[i].ep == ep) {
+			return &usb_dev.transfer[i];
+		}
+	}
+
+	return NULL;
+}
+
+static void usb_transfer_work(struct k_work *item)
+{
+	struct usb_transfer_data *trans;
+	int ret = 0, bytes;
+	u8_t ep;
+
+	trans = CONTAINER_OF(item, struct usb_transfer_data, work);
+	k_sem_take(&trans->sem, K_FOREVER);
+
+	ep = trans->ep;
+
+	if (trans->status != -EBUSY) {
+		/* transfer cancelled or already completed */
+		goto done;
+	}
+
+	if (trans->flags & USB_TRANS_WRITE) {
+		if (!trans->bsize) {
+			if (!(trans->flags & USB_TRANS_NO_ZLP)) {
+				usb_dc_ep_write(ep, NULL, 0, NULL);
+			}
+			trans->status = 0;
+			goto done;
+		}
+
+		ret = usb_dc_ep_write(ep, trans->buffer, trans->bsize, &bytes);
+		if (ret) {
+			/* transfer error */
+			trans->status = -EINVAL;
+			goto done;
+		}
+
+		trans->buffer += bytes;
+		trans->bsize -= bytes;
+		trans->tsize += bytes;
+	} else {
+		ret = usb_dc_ep_read(ep, trans->buffer, trans->bsize, &bytes);
+		if (ret) {
+			/* transfer error */
+			trans->status = -EINVAL;
+			goto done;
+		}
+
+		trans->buffer += bytes;
+		trans->bsize -= bytes;
+		trans->tsize += bytes;
+
+		/* ZLP, short-pkt or buffer full */
+		if (!bytes || (bytes % usb_dc_ep_mps(ep)) || !trans->bsize) {
+			/* transfer complete */
+			trans->status = 0;
+
+			/* some buggy? hardware/sw send short packet with
+			   null byte instead of ZLP, so discard the null byte */
+			if (bytes == 1 && *(trans->buffer -1) == 0) {
```

Comment:
What if the last byte in the transfer is a real '\0'?
In that case, bytes % usb_dc_ep_mps(ep) == 1 and bytes == 1 and we might incorrectly discard it.
If the SW/HW is buggy, is trans->bsize == 0 always? If so, you can add that check in here rather than above.

Reply:
> What if the last byte in the transfer is a real '\0'?
I think this is a gray area, Linux wrongly supposes that hardware is buggy and send a NULL byte instead of a ZLP by default, at least for usbnet (cf http://elixir.free-electrons.com/linux/v4.16-rc1/source/drivers/net/usb/usbnet.c#L1391), so we have to live with it. Maybe I could introduce a flag that class drivers have to set in order to manually discard this byte.